Make the further refinements to CSS property and value code suggested in the review of 256223@main

darinadler opened the pull request:
#### 61d1950d3bd4e746a42c278da6c53ef5ac9af4f3
<pre>
Make the further refinements to CSS property and value code suggested in the review of 256223@main
<a href="https://bugs.webkit.org/show_bug.cgi?id=247495">https://bugs.webkit.org/show_bug.cgi?id=247495</a>
rdar://problem/101967609

Reviewed by Sam Weinig.

* Source/WTF/wtf/FixedVector.h: Added constructor that takes two input iterators.

* Source/WTF/wtf/Vector.h:
(WTF::Mapper::map): Use rvalue reference in the for loop so we can work with iterators that don&apos;t
return actual references: makes sense anyway since we use WTFMove. Also corrected the argument type
in the helper for compactMap to match what&apos;s used in map to avoid an unneeded copy.
(WTF::CompactMapper::compactMap): Ditto.

* Source/WebCore/animation/CSSPropertyAnimation.cpp:
(WebCore::operator&lt;&lt;): Added overload for CSSPropertyID, and use it to simplify logging code below.

* Source/WebCore/css/ComputedStyleExtractor.cpp:
(WebCore::ComputedStyleExtractor::copyProperties): Use compactMap here instead of
reserveInitialCapacity, append, and shrinkToFit.

* Source/WebCore/css/makeprop.pl: Remove unneeded backslashes. Added == member to the range iterator
and size member to the range itself. Added enough specialization of std::iterator_traits to make
compactMap work with the range.
* Source/WebCore/css/makevalues.pl: Ditto.

* Source/WebCore/dom/Document.cpp:
(WebCore::Document::exposedComputedCSSPropertyIDs): Use emplace and FixedVector constructor that
takes input iterators instead of first allocating a FixedVector and then calling std::copy.

Canonical link: <a href="https://commits.webkit.org/256368@main">https://commits.webkit.org/256368@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/69e857a6ae46e91a75d97a7fec3069110d3b258f

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/95572 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/4836 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/28620 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/105152 "Built successfully") | [  ~~🛠 🧪 win~~](https://ews-build.webkit.org/#/builders/10/builds/165422 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/99558 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/4874 "Built successfully") | [✅ 🛠 mac-debug](https://ews-build.webkit.org/#/builders/71/builds/33579 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/87943 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/101003 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/101234 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/78/builds/3562 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/82184 "Built successfully") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/30641 "Passed tests") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/9/builds/85448 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/87361 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/73478 "Passed tests") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/1/builds/86635 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/66/builds/39324 "Built successfully") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/73/builds/18915 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/38/builds/81939 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/67/builds/37024 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/20217 "Passed tests") | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/46/builds/28143 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/4400 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/63/builds/41021 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/60/builds/42864 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/37/builds/84617 "Built successfully") | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/65/builds/43009 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/39468 "Passed tests") | [✅ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/45/builds/19108 "Passed tests") | 
<!--EWS-Status-Bubble-End-->

weinig commented on Source/WTF/wtf/FixedVector.h:
```diff
@@ -61,6 +61,11 @@ class FixedVector {
     }
 
+    template<typename InputIterator> FixedVector(InputIterator begin, InputIterator end)
+        : m_storage(begin == end ? nullptr :Storage::create(begin, end).moveToUniquePtr())
```
There is a missing space after the second `:`.

weinig commented on Source/WTF/wtf/Vector.h:
```diff
@@ -1781,7 +1781,7 @@ struct Mapper {
     using DestinationItemType = typename std::invoke_result<MapFunction, SourceItemType&>::type;
 
-    static Vector<DestinationItemType> map(SourceType source, const MapFunction& mapFunction)
+    static Vector<DestinationItemType> map(const SourceType& source, const MapFunction& mapFunction)
```
Will this cause us to lose the ability to efficiently move during the map? I am thinking of a test case like:

```c++
Vector<MoveOnlyType> source { a, b, c };
auto mapped = WTF::map(WTFMove(source), [] (MoveOnlyType&& item) {
    return TypeThatWrapsMoveOnlyType(WTFMove(item);
});
```

(I will freely admit I don't know if that worked previously).

darinadler replied:
I don’t think it will; I think it will remove unneeded copying with no bad side effects. There’s a separate specialization below with for when the source is an rvalue reference. But I also think it’s not smart for me to modify this without adding tests, so I’ll take these unnecessary changes out of this pull request.

darinadler replied:
Yes, I confirmed there are already move-only tests and this change does not break them. But I broke this change out into a separate pull request, https://github.com/WebKit/WebKit/pull/6178, where I have tests confirming it eliminates unnecessary copies.

weinig commented on Source/WebCore/css/ComputedStyleExtractor.cpp:
```diff
@@ -4345,14 +4345,12 @@ Ref<MutableStyleProperties> ComputedStyleExtractor::copyPropertiesInSet(const CS
+        if (!value)
+            return std::nullopt;
+        return { { property, WTFMove(value) } };
+    }));
```
It's really quite nice :).

darinadler committed:
Make the further refinements to CSS property and value code suggested in the review of 256223@main
https://bugs.webkit.org/show_bug.cgi?id=247495
rdar://problem/101967609

Reviewed by Sam Weinig.

* Source/WTF/wtf/FixedVector.h: Added constructor that takes two input iterators.

* Source/WTF/wtf/Vector.h:
(WTF::Mapper::map): Use rvalue reference in the for loop so we can work with iterators that don't
return actual references: makes sense anyway since we use WTFMove. Also corrected the argument type
in the helper for compactMap to match what's used in map to avoid an unneeded copy.
(WTF::CompactMapper::compactMap): Ditto.

* Source/WebCore/animation/CSSPropertyAnimation.cpp:
(WebCore::operator<<): Added overload for CSSPropertyID, and use it to simplify logging code below.

* Source/WebCore/css/ComputedStyleExtractor.cpp:
(WebCore::ComputedStyleExtractor::copyProperties): Use compactMap here instead of
reserveInitialCapacity, append, and shrinkToFit.

* Source/WebCore/css/makeprop.pl: Remove unneeded backslashes. Added == member to the range iterator
and size member to the range itself. Added enough specialization of std::iterator_traits to make
compactMap work with the range.
* Source/WebCore/css/makevalues.pl: Ditto.

* Source/WebCore/dom/Document.cpp:
(WebCore::Document::exposedComputedCSSPropertyIDs): Use emplace and FixedVector constructor that
takes input iterators instead of first allocating a FixedVector and then calling std::copy.

Canonical link: https://commits.webkit.org/256368@main
